api/src/main/java/me/lokka30/treasury/api/economy/EconomyProvider.java

```diff
@@ -206,54 +314,48 @@ void createAccount(
             }
-            return new HashSet<>();
-        }).thenAccept(identifiers -> {
+            subscription.succeed(ret);
+        });
+    }
+
+    /**
+     * Request all the {@link NonPlayerAccount non-player accounts} where the given player has the given permissions.
+     *
+     * @param playerId     the player
+     * @param permissions  the permissions that the given player has to have on the {@link NonPlayerAccount account}
+     * @see #retrieveAllAccountsPlayerIsMemberOf(UUID)
+     */
+    @NotNull
+    default CompletableFuture<Collection<String>> retrieveAllAccountsPlayerHasPermission(
+            @NotNull UUID playerId, @NotNull AccountPermission @NotNull ... permissions
+    ) {
+        Objects.requireNonNull(playerId, "playerId");
+        Objects.requireNonNull(permissions, "permissions");
+        return this.retrieveAccountIds().thenComposeAsync(identifiers -> {
             if (identifiers.isEmpty()) {
-                subscription.succeed(identifiers);
-                return;
+                return CompletableFuture.completedFuture(Collections.emptySet());
             }
-            Set<String> ret = new HashSet<>();
+            Collection<CompletableFuture<Account>> accounts = new HashSet<>();
             for (String identifier : identifiers) {
-                EconomySubscriber.<Account>asFuture(subscriber -> retrieveAccount(identifier,
-                        subscriber
-                )).exceptionally(throwable -> {
-                    if (throwable instanceof EconomyException) {
-                        subscription.fail((EconomyException) throwable);
-                    } else {
-                        subscription.fail(new EconomyException(EconomyFailureReason.OTHER_FAILURE,
-                                throwable
-                        ));
-                    }
-                    return null;
-                }).thenCompose(account -> {
-                    if (account == null) {
-                        return CompletableFuture.completedFuture(false);
-                    }
-
-                    if (!(account instanceof NonPlayerAccount)) {
-                        return CompletableFuture.completedFuture(false);
-                    }
-                    return EconomySubscriber.asFuture(subscriber -> account.isMember(playerId,
-                            subscriber
-                    ));
-                }).thenAccept(val -> {
-                    if (val) {
-                        ret.add(identifier);
-                    }
-                });
-
+                accounts.add(this.retrieveAccount(identifier));
             }
-            subscription.succeed(ret);
+            return FutureHelper.mjf(
+                    account -> account
+                            .hasPermission(playerId, permissions)
+                            .thenApply(triState -> triState == TriState.TRUE),
+                    Account::getIdentifier,
+                    accounts
+            );
```

Comment:
`FutureHelper#mjf` uses [`CompletableFuture#join()`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#join--) which
> throws an (unchecked) exception if completed exceptionally

but this excepcion is not handled in the `mjf` method or this method, nor is it specifyed in the documentation that it's not handled